### DISCUSSION TITLE
What do you want to see in 6.0 release?

### DISCUSSION THREAD

**Discussion by adrianhall** [Azure/azure-mobile-apps#324] [General]
I've got a list of things I want to do for the 6.0 release of Azure Mobile Apps:

Service side:
1. Support for DateOnly, TimeOnly, and the DbGeography (Point and Area)
2. Support for the DbGeography search terms3. 
3. Support for SignalR transport for real-time delivery of updates.
4. HTTP/2 and potentially QUIC support.

Client side:
1. Retire the .NET Standard support and concentrate on the .NET 6 support
2. ... which allows me to support DateOnly, TimeOnly, and DbGeography capabilities
3. Support for SignalR transport for real-time delivery of updates.
5. Eventing (to allow running code on offline store mutations, for example).
6. Multi-threaded queue submissions (which should 4x the push speed).
7. Pluggable transports (HTTP/2 and QUIC, specifically) - this will also speed up sync operations.
8. Support for a NoSQL local store (like LiteDB) as an alternative to SQLite.

What's on your list?  What would you like to see?

**Comment by fileman** [comment]
8. Using custom id generation in client OfflineTable InsertItemAsync

now we have: `id = Guid.NewGuid().ToString("N");`
proposed: `id = IdCustomGenerator != null ? IdCustomeGenerator.NewId() : Guid.NewGuid().ToString("N");`

where NewId() return always a string, this is an extreme semplification, but I can work on testable code if you like

**Reply by adrianhall** [reply]
I like this idea.  Maybe put a `Func<string, string>? IdGenerator` in the options, then use that if it is available.  Where the parameter is the table name (allowing you to do different generators.  I can then just do `options.IdGenerator?.Invoke(tableName) ?? Guid.NewGuid().ToString("N")`.

Might even get that in 5.0.1 if you want to file the issue.

**Reply by fileman** [reply]
In code and tests the strongly type version of OfflineTable dont check if `T instance` item has id, while the untyped do it for `JObject instance`
Is it intentional or a mistake?

**Reply by adrianhall** [reply]
It's intentional.  When dealing with strongly typed versions, they get handled by the untyped version, so the ID check is done there instead.  It's a case of not duplicating work that you know is done elsewhere.

**Reply by adrianhall** [reply]
... and I should probably do a series of Twitch Q&A sessions on working with the code :-)  Not sure if anyone would be interested, but it would be fun.

**Reply by fileman** [reply]
ok, I reviewed the code when there was calm in the house and after sleeping and now I have more clear how it works.
[here](https://github.com/fileman/azure-mobile-apps/tree/idgenerator) you can find a first draft, I don't know if that's what you thought or do you prefer a PR?

**Reply by adrianhall** [reply]
A PR works best!

**Comment by fileman** [comment]
9. built-in db version check & client side older db clear (see #350)

**Reply by sisaacks** [reply]
@fileman what is this going to do?  Built In DB version check?

**Reply by fileman** [reply]
Your case made me think that having a similar function included might be interesting for others as well.
I will try to implement it but I would like @adrianhall opinion

**Reply by sisaacks** [reply]
@fileman I am trying to do what you suggested using the purge.  My issue comes when I have added a property to the SQL azure using EF code first migration and then adding the property to my model on the client.  The first PUSH causes the error.  Most of my users will be working offline.  So there could be unsynced data, so I dont want to delete the dB file.....Adrian mentioned having two client databases but I am not understanding the logic or how that would work.

**Comment by myokeeh** [comment]
I currently include a "seed" SQLite database with my app updates. This is the same file for offline syncing. It would be nice to have a way to more efficiently download the data from the service for new installs or offline database "rebuilding". Syncing sometimes chokes/times out.

**Comment by adrianhall** [comment]
I've started filing issues in the backlog for the features we want to support.  Please file detailed feature requests for the items you want!

**Comment by fileman** [comment]
10. Async Queryable method in IRepository<T>
`public Task<IQueryable<T>> AsQueryable()`

So we can use custom repository async methods without using Task.Run(...) or .Result

**Reply by adrianhall** [reply]
Any reason you can't do 

``` csharp
var queryable = table.AsAsyncEnumerable().ToEnumerable();
```

The ToEnumerable() method is in System.Linq.Async here: https://github.com/dotnet/reactive/blob/main/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToEnumerable.cs#L22

**Reply by fileman** [reply]
If I understand correctly, ToEnumerable shifts the execution of the query to the client side with all the consequences of the case.

I'm adding Datasync to ABP Solution, so I have to call the async methods of [his repositories](https://docs.abp.io/en/abp/latest/Repositories#querying-linq-over-the-repositories), but it is not excluded that the same problems may arise in different cases.

I can work on this if you like it

**Reply by adrianhall** [reply]
Ah - server side logic, not client side.  Got ya!

Small change to do

``` csharp
public Task<IQueryable<T>> AsQueryableAsync() => Task.FromResult(AsQueryable()); 

public IQueryable<T> AsQueryable() { /* whatever */ }
```

Then the service calls AsQueryableAsync(), and that wraps to the AsQueryable() automatically.  if you need the async version, then you can override it.  I like this idea.
